src/python/datastore/data_types.py

```diff
@@ -631,6 +631,10 @@ def _pre_put_hook(self):
     self.populate_indices()
 
   def _post_put_hook(self, _):
+    if not self.key:
+      # Failed put.
+      return
```

Comment:
Do we need to log ? How is this expected, maybe expand comment.

Reply:
This was always wrong -- _post_put_hook can be called even if there was an exception in put(), and the exception will be rethrown automatically afterwards. Expanded the comment a bit more.